### ISSUE TITLE
Implement better parent-child relationship scheme for target installation

### ISSUE THREAD

**Issue by ar13pit** [tue-robotics/tue-env#332]
The current scheme is simple but weak. What I mean by this is lets say we install `tue-dev-full`. In the list of installed packages a file named `tue-dev-full` appears. This target depends on `tue-dev` and `tue-tools`. `tue-dev` installs `ros-action_server` as it is one of its dependencies. Now we proceed towards the installation of `ros-action_server`. A new file in the installed packages directory appears with the name `ros-action_server`.

What I propose here:
Get rid of `dependencies` and `dependencies-on` directories and have only one directory named `installed`. Lets say we install `tue-dev-full`. First a directory named `tue-dev-full` is to be created in `installed` and it must contain two files or directories named as `parent` and `child`. At the moment it has no `child` so this can be omitted. In the parent file/directory, there will be a list of all its parent dependencies created recursively. And then each of this parent dependency will be installed in the similar way as `tue-dev-full`. This way when installing `ros-action_server` after `tue-dev-full`, we can be sure that it is already installed by just looking at one directory instead of trying to decipher the `dependencies` and `dependencies-on` directories. With this another benefit that exists is a target which has no children can be safely deleted from workspace during removal and cleaning up of the workspace becomes easy.

**Comment by MatthijsBurgh** [comment]
The suggested scheme is very close to the current scheme. Instead of having a directory with a lot of targets. Which each target having a parent and child file/directory. We now have a dependency and dependency-on directories, which are similar to parent and child in the suggested scheme, with a lot of targets. So it the new scheme would just be a switch of directory levels.

How we interact with this `database` could be improved.

**Comment by ar13pit** [comment]
True. The `env` toolkit would become more intelligent than it currently is.

Maybe we could use something like SQLite to make an elegant `database` because it is a standard practice to use such a thing for storing information like this. It has some added benefits. We can have a `tue-cd` command to navigate to any package built from source. Also get a list of all installed packages so that the case I mentioned above doesn't occur and also be able to generate a list of orphaned packages.

**Comment by MatthijsBurgh** [comment]
If we would be refactoring this much, we might consider switching to [ros-get](github.com/rayman/ros-get), but that doesn't have priority at the moment.

**Comment by ar13pit** [comment]
That would be a big jump and to be honest I don't want to get rid of bash. In the suggested scheme we would still keep most of what we have right now. I'll open a parallel branch to the master and start small as in my opinion most changes will only be made to the `tue-install-target` function in `tue-install-impl.bash` and some functions will get added to `tue-get`.

**Comment by MatthijsBurgh** [comment]
Maybe put this on hold till speech recognition in working. We need to focus. This has been distracting me as well.

**Comment by ar13pit** [comment]
This is not a priority in any scenario.
